### PR TITLE
[BRO] Implemented Teferi, Temporal Pilgrim

### PR DESCRIPTION
**Pull request by weirddan455**

### CODE REVIEW

File: Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        if (opponent == null) {
+            return false;
+        }
+        TargetControlledPermanent target = new TargetControlledPermanent();
+        target.setNotTarget(true);
+        opponent.chooseTarget(Outcome.ReturnToHand, target, source, game);
+        Permanent toHand = game.getPermanent(target.getFirstTarget());
+        if (toHand != null) {
+            opponent.moveCards(toHand, Zone.HAND, source, game);
+        }
```
Comment (Alex-Vasile) [inline Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:91]: Is this correct for a card they control but not own?

Comment (weirddan455) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:91]: Yea, that's all `ReturnToHandEffect` does.  The engine won't let you have other player's cards in your hand as far as I know.

File: Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        Player opponent = game.getPlayer(source.getFirstTarget());
+        if (opponent == null) {
+            return false;
+        }
+        TargetControlledPermanent target = new TargetControlledPermanent();
+        target.setNotTarget(true);
+        opponent.chooseTarget(Outcome.ReturnToHand, target, source, game);
+        Permanent toHand = game.getPermanent(target.getFirstTarget());
+        if (toHand != null) {
+            opponent.moveCards(toHand, Zone.HAND, source, game);
```
Comment (Alex-Vasile) [inline Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: I believe you should be applying the first part of the effect before moving on since it's worded as "Do x. THEN do y". The other cards that get shuffled into the library should see the first one be returned to hand.

Comment (weirddan455) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: I'm not sure what you mean by that.

Comment (Alex-Vasile) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: Sorry, let me try again. The -12 effect has 2 parts. The first moves a permanent from the battlefield to the hand. The second effect more the other permanents to their owner's libraries.

Because of how the effect is worded they should be happening sequentially so that the permanent returned to the hand is already in the hand and can't see any of the other permanents leave the battlefield.

Though I am less sure if that's an issue here if you are calling the moveCard functions directly.

Comment (weirddan455) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: Yes, I understand that.  I don't know what you're asking me to change in the code.

Comment (Alex-Vasile) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: Had to look into it a little (see https://github.com/magefree/mage/pull/9304#discussion_r927718153)

Call `game.applyEffects()`

Comment (awjackson) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: That should be game.getState().processAction().

Comment (weirddan455) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: Added `game.getState().processAction()` in latest commit.  Also changed the owner HashSet to store UUID instead of Player.  Probably doesn't matter too much but I think it should be easier to hash and less likely to have bugs (Player concedes mid resolution or something.)

Comment (Alex-Vasile) [reply Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java:90]: > Added `game.getState().processAction()` in latest commit. Also changed the owner HashSet to store UUID instead of Player. Probably doesn't matter too much but I think it should be easier to hash and less likely to have bugs (Player concedes mid resolution or something.)

Yes that's also a good idea! I missed that issue.
